File: leetcode_py/p416.py
from typing import List


class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        if len(nums) < 2:
            return False

        s = sum(nums)
        maxnum = max(nums)
        if s % 2 != 0 or maxnum > s // 2:
            return False

        dp = [False] * (s // 2 + 1)

        dp[0] = True
        for i in range(1, len(nums)):
            for j in range(len(dp) - 1, nums[i], -1):
                dp[j] = dp[j] or dp[j - nums[i]]

        return dp[-1]

File: leetcode_C++/p416.cpp
/*
@Author: mrxhar
@Date:   2023-08-29 10:25:25
@Email:  1337904145@qq.com
@Blog:   https://leetcode.cn/problems/partition-equal-subset-sum/
*/
#include <vector>
#include <algorithm>
#include <numeric>

using namespace std;

class Solution
{
public:
    bool canPartition(vector<int> &nums)
    {
        int n = nums.size();
        // 和如果不是偶数肯定是不行的，我们也可也使用双指针
        if (n < 2)
        {
            return false;
        }

        int sum = accumulate(nums.begin(), nums.end(), 0);
        int maxNum = *max_element(nums.begin(), nums.end());
        if (sum & 1)
        {
            return false;
        }
        int target = sum / 2;
        if (target < maxNum)
        {
            return false;
        }

        vector<int> dp(target + 1, 0);
        dp[0] = true;

        for (int i = 1; i < n; i++)
        {
            int num = nums[i];
            for (int j = target; j >= num; j--)
            {
                dp[j] |= dp[j - num];
            }
        }
        return dp[target];
    }
};